gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -109,12 +132,23 @@ def self.signature_for_method(method)
   # Only public because it needs to get called below inside the replace_method blocks below.
   def self._on_method_added(hook_mod, method_name, is_singleton_method: false)
     current_declaration = T::Private::DeclState.current.active_declaration
-    return if !current_declaration
-    T::Private::DeclState.current.reset!
-
     mod = is_singleton_method ? hook_mod.singleton_class : hook_mod
     original_method = mod.instance_method(method_name)
 
+    if final_method?(original_method)
+      raise "`#{mod.name}##{method_name}` was declared as final and cannot be redefined"
+    end
+    mod.ancestors.each do |a|
+      (a.instance_methods(false) + a.private_instance_methods(false)).each do |m|
+        if m == method_name && final_method?(a.instance_method(method_name))
+          raise "`#{a.name}##{method_name}` was declared as final and cannot be overridden in `#{mod.name}`"
```

Comment:
Did you think about whether we should use one of the handlers in `T::Configuration` instead of `raise`ing directly here?

tbc I think we shouldn't use T::Configuration here until someone asks for it, but curious if you had thoughts.

Reply:
Unless someone asks for something else I think just raising here is fine. We do similar things in the abstract checks.